### ISSUE TITLE
Rewrite of CI to use the GHCR to store intermediate build artifacts

### ISSUE THREAD

**Issue by muthuArivoli** [DukeRobotics/robosub-ros#215]
## Overview
A rewrite of our docker autobuild system to use the new [GitHub container registry](https://docs.github.com/en/packages/getting-started-with-github-container-registry/about-github-container-registry) (GHCR) to store intermediate images generated during PRs, rather than using Github artifacts. We will also then use docker's official [buildx](https://github.com/docker/setup-buildx-action) and [qemu](https://github.com/docker/setup-qemu-action) actions.

## Background
GitHub actions does not really allow us to start containers in the middle of a job; all jobs must be predefined to run inside a container: either the image is defined by us, or is the default one that GitHub uses. This wouldn't be a problem, but unfortunately, the image must be located in a public registry, which means we cannot use local images for the containers. We want to use the image that we just built to test our code in during PRs, however since the image we just built is still only local, we currently "trick" GitHub actions by using docker compose and recloning the repo into a container created from the local image. This will fail on forks as well as PRs that have a target branch that is something other than master as noted in #212, thus it is not robust enough to handle all situations.

Also, our current docker images are structured so that they are dependent on each other (onboard and landside depend on core). The forces us to save the core image that is created, then reload that image during the next image's build stage, since otherwise PRs would not generate accurate images (see DukeRobotics/robosub-ros-docker#33). Due to this, we spend time during each run saving the core image into an archive, and uploading it as a GitHub artifact, after which we download and load it into the next stage. This also forces us to use the very slow method of qemu and the docker build command, even though buildx was designed for our need of multi-platform images. This is due to buildx being unable to read local images when building. Overall, our current method is quite slow, which could cause problems like races if people merge code into master in quick succession.

## Proposal

Thus, to make our CI more robust, I propose that we store all images we generate in PRs on the GHCR, for the duration of the run. This serves as a replacement to saving the images as GitHub artifacts. This will allow us to use GitHub actions as it's intended for - we can define the image to test with as the one we just pushed to the GHCR. Doing this will also allow us to use buildx, since the images are now in GHCR, which buildx can read from to generate accurate images during PRs. To create unique images for each PR/commit, we can use the commit's SHA as part of the image tag. We can then just push the image to the GHCR, and later jobs will always have access to the commit SHA, so they will know what image to pull. For pushes to master, we would just use the main tags like we do now, and push to docker hub instead.

We can't just use docker hub since they have a limit on pulls/pushes (200 per 6 hours), which we could easily run out of (that would mean a max of about 20-30 commits per 6 hours). The GHCR poses no restrictions (from what I could tell) so it would be advantageous to not have to deal with that. Also, we might see some nice integration between the GHCR and GitHub Actions in the future. 

## Pros

This will allow us to remove most, if not all, of the bash scripts that will be located in the `.github/workflows` directory once #212 is merged. It would make the workflow more clear, since we'll be using more of the official actions that docker provides, and people will not have to decipher our build and setup scripts, and won't have to understand how to setup the multiarch builds. We will also probably see a speedup in our builds, especially since we can use buildx. Future PRs can use the caching feature in the docker actions/buildx to speed up builds even more. 

## Cons
GHCR is still currently in beta, so we might see some problems initially with integration. This would also take up time.

cc @neildhar

**Comment by muthuArivoli** [comment]
A few updates on this - apparently the GHCR, and github packages in general, does not allow you to delete public packages. So, we won't be able to store our temp images as public packages, and since private packages cost money above 500 MB, it seems that we will not be able to use the GHCR to do stuff. However, I have realized that we can fix the issue with branches not using the appropriate base branch with PRs that are not targeting the master branch, so I'll either open a new issue for that, or just open a PR that will fix that.

**Comment by muthuArivoli** [comment]
Closing this in favor of #234, will revisit this if the GHCR ever allows deletion of public images.
